src/sha3.js

```diff
@@ -9,44 +11,53 @@ const functions = [
   [0x17, sha3.sha3_224],
   [0x18, sha3.shake128, 256],
   [0x19, sha3.shake256, 512],
-  [0x1A, sha3.keccak224],
-  [0x1B, sha3.keccak256],
-  [0x1C, sha3.keccak384],
-  [0x1D, sha3.keccak512]
+  [0x1a, sha3.keccak224],
+  [0x1b, sha3.keccak256],
+  [0x1c, sha3.keccak384],
+  [0x1d, sha3.keccak512]
 ]
 
-class Hasher {
-  constructor (hashFunc, arg) {
+type HexString = string
+type Sha3Hash = Buffer
+type ShaHasher = (input: string | Buffer, length?: number) => HexString
+
+class ShaHash implements Hash {
+  hf: ShaHasher
+  input: Buffer | null
+  arg: number
+
+  constructor(hashFunc, arg?: number) {
     this.hf = hashFunc
-    this.arg = arg
+    if (arg) {
+      this.arg = arg
+    }
     this.input = null
   }
 
-  update (buf) {
+  update(buf: Buffer): Hash {
     this.input = buf
     return this
   }
 
-  digest () {
+  digest(): Sha3Hash {
+    if (!this.input) {
+      throw Error("Missing an input to hash")
+    }
```

Comment:
In general I would really recommend encoding such invariants in a type system as primary benefit of type checking is to eliminate possibility of runtime errors. I don't know exact specifics of how this is used so my specific suggestion may not necessarily be applicable but it might clarify what I mean. Usually this can be solved by types where you can't have an invalid state. Here is an example:

```js
interface HashUpdate {
  update(buf: Buffer): Hash;
}

interface Hash extends HashUpdate {
  digest(): Buffer;
}


class ShaHash implements Hash {
  // Assumbtion is that this static function is exposed and not the class so that consumer can only create
  // HashUpdate and can't create `Hash` without an input. Since our `HashUpdate` interface does not have
  // a `digest` attempt to call will be reported by a type checker.
  static new(hashFunc, arg?: number):HashUpdate {
    return new ShaHash(hashFunc, arg)
  }
  // Since `update` returns a `Hash` that has both `digest` and `update` methods
  // calling `digest` on returned value will be by fine by type checker. Please also note
  // that even though it's the same instance from type checker perspective former still
  // has not `update` and later does.
  update(buf: Buffer): Hash {
    this.input = buf
    return this
  }
  // It is necessary to check this as `input` can be `null` but either way it's useful to cover
  // consumers that don't use type-checker
  digest(): Sha3Hash {
    if (!this.input) {
      throw Error("Missing an input to hash")
    }
    //...
  }
}
```